From 6f01c97d8e6d43f976a4a8b2757d59e7a74f4c9d Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Fri, 19 May 2023 15:49:58 +0800
Subject: [PATCH 026/263] [Huawei] userswap: fix BUG_ON in __mcopy_atomic()

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

When the swap-in buffer contains no physical pages, the errno in
mfill_atomic_pte_nocopy() will be ENOENT. A BUG_ON error will occur
because the userswap feature does not use the struct page *page and page
is set to NULL.
To fix this issue, the errno should be changed from ENOENT to EINVAL.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/userswap.c b/mm/userswap.c
index 2d47f6ed9f91..7a0c23e5ed59 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -407,7 +407,7 @@ int mfill_atomic_pte_nocopy(struct mm_struct *mm,
 
 	src_vma = find_vma(mm, src_addr);
 	if (!src_vma || src_addr < src_vma->vm_start)
-		return -ENOENT;
+		return -EINVAL;
 
 	if (src_vma->vm_flags & VM_LOCKED)
 		return -EINVAL;
-- 
2.17.1

